stl/inc/algorithm

```diff
@@ -2556,6 +2580,253 @@ namespace ranges {
+        _NODISCARD constexpr auto operator()(_It _First, _Se _Last, _Ty _Init, _Fn _Func) const {
+            _Adl_verify_range(_First, _Last);
+            return _Fold_left_with_iter_impl<_It>(
+                _STD move(_First), _STD move(_Last), _STD move(_Init), _Pass_fn(_Func));
+        }
+
+        template <input_range _Rng, class _Ty, _Indirectly_binary_left_foldable<_Ty, iterator_t<_Rng>> _Fn>
+        _NODISCARD constexpr auto operator()(_Rng&& _Range, _Ty _Init, _Fn _Func) const {
+            return _Fold_left_with_iter_impl<borrowed_iterator_t<_Rng>>(
+                _RANGES begin(_Range), _RANGES end(_Range), _STD move(_Init), _Pass_fn(_Func));
+        }
+
+    private:
+        template <class _RetIt, class _It, class _Se, class _Ty, class _Fn>
+        _NODISCARD constexpr auto _Fold_left_with_iter_impl(_It&& _First, _Se&& _Last, _Ty&& _Init, _Fn _Func) const {
+            _STL_INTERNAL_STATIC_ASSERT(input_iterator<_It>);
+            _STL_INTERNAL_STATIC_ASSERT(sentinel_for<_Se, _It>);
+            _STL_INTERNAL_STATIC_ASSERT(_Indirectly_binary_left_foldable<_Fn, _Ty, _It>);
+
+            using _Uty         = decay_t<invoke_result_t<_Fn&, _Ty, iter_reference_t<_It>>>;
+            using _Return_type = fold_left_with_iter_result<_RetIt, _Uty>;
+
+            if (_First == _Last) {
+                return _Return_type{_STD move(_First), static_cast<_Uty>(_STD move(_Init))};
+            } else {
+                auto _UFirst = _Unwrap_iter<_Se>(_STD move(_First));
+                auto _ULast  = _Unwrap_sent<_It>(_STD move(_Last));
+
+                _Uty _Accum = _STD invoke(_Func, _STD move(_Init), *_UFirst);
+                for (++_UFirst; _UFirst != _ULast; ++_UFirst) {
+                    _Accum = _STD invoke(_Func, _STD move(_Accum), *_UFirst);
+                }
+
+                _Seek_wrapped(_First, _STD move(_UFirst));
+                return _Return_type{_STD move(_First), _STD move(_Accum)};
+            }
+        }
+    };
+
+    _EXPORT_STD inline constexpr _Fold_left_with_iter_fn fold_left_with_iter{_Not_quite_object::_Construct_tag{}};
+
+    class _Fold_left_fn : private _Not_quite_object {
+    public:
+        using _Not_quite_object::_Not_quite_object;
+
+        template <input_iterator _It, sentinel_for<_It> _Se, class _Ty, _Indirectly_binary_left_foldable<_Ty, _It> _Fn>
+        _NODISCARD constexpr auto operator()(_It _First, _Se _Last, _Ty _Init, _Fn _Func) const {
+            return _RANGES fold_left_with_iter(_STD move(_First), _Last, _STD move(_Init), _Pass_fn(_Func)).value;
+        }
+
+        template <input_range _Rng, class _Ty, _Indirectly_binary_left_foldable<_Ty, iterator_t<_Rng>> _Fn>
+        _NODISCARD constexpr auto operator()(_Rng&& _Range, _Ty _Init, _Fn _Func) const {
+            return _RANGES fold_left_with_iter(_STD forward<_Rng>(_Range), _STD move(_Init), _Pass_fn(_Func)).value;

```

Comment:
Ideally we'd want this to avoid the extra move into the `in_value_result` and back out. (This is why the wording uses _Returns:_ instead of "_Effects:_ Equivalent to `return ...;`")

Reply:
Do RVO or deferred temporary materialization help with that?

I think this would be reasonable to address in a followup PR.